build(docs-infra): several updates/improvements for the PR preview server

gkalpak opened the pull request:
See individual commits for details.

petebacondarwin commented:
Tiny typo in the 3rd commit: `This helps surface such error.` -> `This helps surface such errors.`

gkalpak commented:
Thx, fixed the typo.

mary-poppins commented:
You can preview bb3e2bf at https://pr36837-bb3e2bf.ngbuilds.io/.

petebacondarwin commented on aio/aio-builds-setup/dockerbuild/scripts-sh/health-check.sh:
```diff
@@ -8,10 +8,32 @@ exitCode=0
+    return
+  fi
+
+  openssl x509 -checkend 0 -in "$certPath" -noout > /dev/null
```
As I understand it `-checkend X` is checking whether the cert expires in the next X seconds. If so, is it worth giving us some notice, like setting X to something larger like 1 hour?

gkalpak replied:
This is mainly for detecting outdated, locally generated, self-signed certs. Theoretically, they are generated on every `docker build`, but Docker might use an intermediate image layer from the cache and not run the command.

And on production, the health-check is not the best way to get notified about soon-to-expire certs :grin:
Not to mention that we are currently not monitoring the health-check result in real-time.

Bottom line: I don't think this would provide any benefit as things are set up currently.

petebacondarwin commented on aio/aio-builds-setup/dockerbuild/Dockerfile:
```diff
@@ -79,13 +79,18 @@ RUN echo "deb https://dl.yarnpkg.com/debian/ stable main" | tee /etc/apt/sources
 
 # Install packages
+# NOTE: Some packages (such as `nginx`, `nodejs`, `openssl`) make older versions unavailable on the
+#       repositories, so we cannot pin to specific versions for these packages :(
```
Strictly we can pin them but we would need to update them in a timely fashion. By not pinning we are relaxing the versions in order to avoid unwanted outages where we have not updated the versions soon enough. Right?

gkalpak replied:
Right.

petebacondarwin commented on aio/aio-builds-setup/scripts/update-preview-server.sh:
```diff
@@ -7,9 +7,9 @@ echo -e "\n\n[`date`] - Updating the preview server..."
-readonly HOST_BUILDS_DIR=$4
+readonly HOST_SECRETS_DIR=$2
+readonly HOST_BUILDS_DIR=$3
+readonly HOST_LOCALCERTS_DIR=$4
```
What? Not alphabetical??

gkalpak replied:
I wondered about that too. It's very not-like-me to have these in another order, right? 🤔
I guess one reason is that [the last three][1] are optional, so it wouldn't make sense to have them interleaved with the mandatory ones.

[1]: https://github.com/angular/angular/blame/1d9375d115b1799cd7335a0b410406b84c8ee4cb/aio/aio-builds-setup/docs/vm-setup--start-docker-container.md#L58-L71

petebacondarwin reviewed (approved):
A few nits. Nice clean up

AndrewKushnir commented:
@gkalpak FYI I've removed "merge" label for now and added "cleanup", since there are some comments from Pete. Please put the "merge" label back when this PR is ready to be merged. Thank you.

petebacondarwin commented:
Oops! Sorry I forgot to add the cleanup label.

gkalpak commented on aio/aio-builds-setup/dockerbuild/scripts-js/test/common/github-pull-requests.spec.ts:
```diff
@@ -47,7 +47,7 @@ describe('GithubPullRequests', () => {
 
     it('should make a POST request to Github with the correct pathname, params and data', () => {
-      githubApi.post.and.callFake(() => Promise.resolve());
+      githubApi.post.and.callFake(() => Promise.resolve() as Promise<any>);
```
Oh, I didn't think of it (latest JAsmine is new to me :grin:).
I'll give it a go. Thx!

gkalpak replied:
Added a fixup commit. Worked like a charm 🤩

mary-poppins commented:
You can preview 8c2daac at https://pr36837-8c2daac.ngbuilds.io/.

petebacondarwin commented:
There are a few more places that could use `resolveTo()` and also there are some places that could use `rejectTo()`...

mary-poppins commented:
You can preview e37ba8d at https://pr36837-e37ba8d.ngbuilds.io/.

gkalpak committed:
build(docs-infra): upgrade all JS dependencies to latest versions

This commit upgrades all dependencies in `scripts-js/` to latest
versions and also includes all necessary code changes to ensure the
tests are passing with the new dependency versions.

gkalpak committed:
fixup! build(docs-infra): upgrade all JS dependencies to latest versions

gkalpak committed:
build(docs-infra): upgrade preview server Docker image to Debian 10 (buster)

Previously, the preview server Docker image was based on Debian 9
(stretch).

This commit upgrades the preview server Docker image to Debian 10
(buster) and also upgrades all dependencies to latest versions
(including upgrading Node.js from v10 to v12).

(The GCE VM running the preview server Docker container was also
upgraded from Debian 9 to 10 on 2020-04-27.)

---
Other changes:
- Pinned the installed version of `curl` to make the `aio-health-check`
  and `aio-verify-setup` checks (which use `curl`) more deterministic.
- Dropped the `*-backports` Debian repositories, since they are no
  longer needed. The `*-backports` repositories were introduced in
  593fe5ed259d5b118d41e2b961b462e922762700 to install `nginx` from, but
  became obsolete in 2f1a862b831f44b9b923dcc52f94af69d3b40aa6, which
  switched to installing `nginx` from the regular repositories again.
- Added `vim` to the list of installed dependencies (for convenience
  during debugging).

gkalpak committed:
test(docs-infra): check TLS certificates as part of preview server's health check

In order to ease local development, self-signed SSL/TLS certificates are
created when building the preview server Docker image. These
certificates are valid for 365 days. Thus, it is possible for an old
certificate to be re-used past its expiration date due to Docker's
caching intermediate layers.

Previously, this would lead to hard-to-debug failures in the
`aio-health-check` and `aio-verify-setup` checks. Even after finding out
that the failures were caused by an expired certificate, it was not
obvious why that would be the case.

This commit adds an additional check to the `aio-health-check` command
that checks the certificates' expiration dates. This helps surface such
errors. It also prints a more helpful message, prompting the user to
build the Docker image with the `--no-cache` option to fix the problem
with self-signed certificates.

gkalpak committed:
build(docs-infra): unpin certain dependencies of the preview server

Previously, in order to remain as deterministic as possible, the
Dockerfile for the preview server Docker image had all dependencies
pinned to specific versions. It turns out that some packages (such as
`nginx`, `nodejs`, and `openssl` - potentially others too) make older
versions unavailable on the repositories once a newer version is
available.

See for example:
- https://github.com/nodesource/distributions/issues/33
- https://askubuntu.com/questions/715104/how-can-i-downgrade-openssl-via-apt-get

This commit, therefore, removes the exact versions for these packages.
The latest versions will be installed everytime the Docker image is
built (subject to Docker caching).

gkalpak committed:
docs(docs-infra): update preview server setup instructions

I recently went through the process of setting up a preview server VM
again and updated the instructions and references based on the latest
docs for Debian, Docker, Google Compute Engine, etc.

gkalpak committed:
refactor(docs-infra): update argument order for `update-preview-server.sh` script

Update the order in which the `update-preview-server.sh` script expects
its arguments (and the associated docs) to be consistent with the order
of arguments in other commands/docs (such as
`vm-setup--start-docker-container.md`).

gkalpak committed:
build(docs-infra): remove linting from the preview server `dev` npm script

Previously, the `dev` npm script in `aio/aio-builds-setup/scripts-js/`
(the PR preview server implementation) would run both linting and unit
tests. This was slow and delayed the feedback loop on each change.

Since the `dev` script should be run during development and give
feedback as fast as possible, this commit removes the linting from the
`dev` script and only keeps the unit tests. Linting is still run in the
`test` npm script (which is more comprehensive). Also, in most cases the
developer's IDE will show linting errors in real time in the editor.

gkalpak committed:
refactor(docs-infra): take advantage of latest Jasmine features in preview server tests

This commit updates the preview server tests to take advantage of
features supported in the latest version of Jasmine that were not
supported when the rests were first written.

Changes include:

- Use [async/await] in tests.
- Use the new [toBeInstanceOf()] and [toHaveBeenCalledBefore()] matchers.
- Use the new [toBeResolved()] and [toBeRejected()] async matchers (and
  their variants).
- Use the new [withArgs()] method of `Spy` to simplify "trained"
  responses.
- Use the new [resolveTo()]/[rejectWith()] methods of `SpyStrategy` (and
  their variants) to simplify promise-based spies.
- Implement custom async matchers (via [addAsyncMatchers()]) to simplify
  certain tests.

[addAsyncMatchers()]: https://jasmine.github.io/api/3.5/jasmine.html#.addAsyncMatchers
[async/await]: https://jasmine.github.io/tutorials/async
[rejectWith()]: https://jasmine.github.io/api/3.5/SpyStrategy.html#rejectWith
[resolveTo()]: https://jasmine.github.io/api/3.5/SpyStrategy.html#resolveTo
[toBeInstanceOf()]: https://jasmine.github.io/api/3.5/matchers.html#toBeInstanceOf
[toBeRejected()]: https://jasmine.github.io/api/3.5/async-matchers.html#toBeRejected
[toBeResolved()]: https://jasmine.github.io/api/3.5/async-matchers.html#toBeResolved
[toHaveBeenCalledBefore()]: https://jasmine.github.io/api/3.5/matchers.html#toHaveBeenCalledBefore
[withArgs()]: https://jasmine.github.io/api/3.5/Spy.html#withArgs

gkalpak committed:
fix(docs-infra): exit with an error when cleaning up the preview server fails

Previously, when the preview server `build-cleanup` script failed, the
error was logged but not reflected to the commands exit code. This seems
to have been accidentally broken in #23576.

This commit fixes it by ensuring the error is re-thrown from the
`BuildCleaner#cleanUp()` method to allow the process to exit with an
error exit code.

gkalpak committed:
test(docs-infra): remove incorrect preview server test

The test was introduced in #23576, but the behavior the test was
verifying does not match the actual behavior of
`BuildCleaner#getOpenPrNumbers()`.

The reason that the test did not fail is that the verification happened
asynchronously, but the test completed synchronously (by accident).

gkalpak committed:
test(docs-infra): enable accidentally disabled preview server tests

Previously, some preview server tests were only running for public
builds. In the past, these tests were run for both public and non-public
builds. The non-public builds tests were disabled in #23576, probably
during debugging some failure.

This commit fixes it by ensuring the tests are run for both public and
non-public builds.

mary-poppins commented:
You can preview 845036a at https://pr36837-845036a.ngbuilds.io/.

gkalpak commented:
Added 5 more commits (at the end), mainly "modernizing" all preview server tests (and some other small fixes):
- **build(docs-infra): remove linting from the preview server `dev` npm script** (dd69303d424fcaefde32455fa55bd4f347e2a29b)
- **fix(docs-infra): exit with an error when cleaning up the preview server fails** (101a2b242e97997453d10a80c57241d48d2bc18a)
- **test(docs-infra): remove incorrect preview server test** (6814f7aca186cc4289597e965f66f8c43c297b88)
- **test(docs-infra): enable accidentally disabled preview server tests** (f928bde1d04e65b64145f5b1c4a3994e784af398)
- **refactor(docs-infra): take advantage of latest Jasmine features in preview server tests** (845036a7d807781336d6c760091751ece5ae2fea)

@petebacondarwin, ptal 😃

petebacondarwin commented on aio/aio-builds-setup/dockerbuild/scripts-js/test/preview-server/build-creator.spec.ts:
```diff
@@ -67,178 +69,152 @@ describe('BuildCreator', () => {
-        expect(promise).toEqual(jasmine.any(Promise));
+        // Do not complete the test (and release the spies) synchronously to avoid running the actual
+        // `extractArchive()`.
+        await promise;
```
If we don't care for the promise to complete then perhaps this test should not be `async`? E.g. 

```ts
it('should return a promise', () => {
  const promise = bc.create(pr, sha, archive, isPublic);
  expect(promise).toBeInstanceOf(Promise);
});
```

gkalpak replied:
As mentioned in the comment, we should _not_ let the test complete before the promise completes, because what would release the spies and we might interact with actual implementations (which is not desirable).

So, while we don't indeed care about the resolved value, we need to wait out the promise.

petebacondarwin commented on aio/aio-builds-setup/dockerbuild/scripts-js/test/preview-server/build-creator.spec.ts:
```diff
@@ -517,78 +480,68 @@ describe('BuildCreator', () => {
-        });
+      it('should abort and reject if it fails to remove the build artifact file', async () => {
+        // tslint:disable-next-line: no-string-throw
+        shellRmSpy.and.callFake(() => { throw 'Test'; });
```
Could this be?

```suggestion
        shellRmSpy.and.throwError('Test');
```

gkalpak replied:
It could, but that is slightly different. `.and.throwError('Test')` is equivalent to `.and.callFake(() => { throw new Error('Test'); })`.

petebacondarwin commented on aio/aio-builds-setup/dockerbuild/scripts-js/test/preview-server/build-creator.spec.ts:
```diff
@@ -609,85 +562,72 @@ describe('BuildCreator', () => {
 
-      expect(promise).toEqual(jasmine.any(Promise));
+      // Do not complete the test (and release the spies) synchronously to avoid running the actual `ls()`.
+      await promise;
```
See other similar test comment.

gkalpak replied:
See https://github.com/angular/angular/pull/36837#discussion_r419525399.

gkalpak commented on aio/aio-builds-setup/dockerbuild/scripts-js/test/common/github-teams.spec.ts:
```diff
@@ -16,6 +16,7 @@ describe('GithubTeams', () => {
       expect(() => new GithubTeams(githubApi, '')).
         toThrowError('Missing or empty required parameter \'githubOrg\'!');
     });
+
```
Consistency :grin:
